proposals/type-classes.md

```diff
@@ -0,0 +1,298 @@
+  fun Service.config(): Config = ProdConfig
+}
+```
+
+```kotlin
+package test
+
+object TestContext: Context<Service> {
+  fun Service.config(): Config = TestConfig
+}
+```
+
+```kotlin
+package prod
+
+service.config() // ProdConfig
+```
+
+```kotlin
+package test
+
+service.config() // TestConfig
+```
+
+## Overcoming `inline` + `reified` limitations
+
+Type classes allow us to workaround `inline` `reified` generics and their limitations and express those as type classes instead:
+
+```kotlin
+interface Reified<A> {
+    val selfClass: KClass<A>
+}
+```
+
+Now a function that was doing something like:
+
+```kotlin
+inline fun <reified A> foo() { .... A::class ... }
+```
+
+can be replaced with:
+
+```kotlin
+fun <A> fooTC(): Klass<A> given Reified<A> { .... A.selfClass ... }
+```
+
+This allows us to obtain generics info without the need to declare the functions `inline` or `reified` overcoming the current limitations of inline reified functions that can't be invoked unless made concrete from non reified contexts.
+
+```kotlin
+class Foo<A> {
+   val someKlazz = foo<A>() //won't compile because class disallow reified type args.
+}
+```
+
+```kotlin
+class Foo<A> {
+   val someKlazz = fooTC<A>() //works anddoes not requires to be inside an `inline reified` context.
+}
+```
+
+## Composition and chain of evidences
+
+Type class instances and declarations can encode further constrains in their generic args so they can be composed nicely:
+
+```kotlin
+package optionext
+
+class OptionMonoid<A> : Monoid<Option<A>> given Monoid<A> {
+
+  fun empty(): Option<A> = None
+
+  fun Option.combine(ob: Option<A>): Option<A> =
+    when (this) {
+      is Some<A> -> when (ob) {
+                      is Some<A> -> Some(this.value.combine(b.value)) //works because there is evidence of a Monoid<A>
+                      is None -> ob
+                    }
+      is None -> this
+  }
+    
+}
+```
+
+The above instance declares a `Monoid<Option<A>>` as long as there is a `Monoid<A>` in scope.
+
+```kotlin
+import optionext.OptionMonoid
+import intext.IntMonoid
+
+Option(1).combine(Option(1)) // Option(2)
+Option("a").combine(Option("b")) // does not compile. Found `Monoid<Option<A>>` instance providing `combine` but no `Monoid<String>` instance was in scope
+```
+
+We believe the above proposed encoding fits nicely with Kotlin's philosophy of extensions and will reduce the boilerplate compared to other langs that also support typeclasses such as Scala where this is done via implicits.
+
+## Typeclasses over type constructors
+
+We recommend if this proposal is accepted that a lightweight version of higher kinds support is included to unveil the true power of typeclasses through the extensions mechanisms
+
+A syntax that would allow for higher kinds in these definitions may look like this:
+
+```kotlin
+interface FunctionK<F<_>, G<_>> {
+  fun <A> invoke(fa: F<A>): G<A>
+}
+
+object Option2List : FunctionK<Option, List> {
```

Comment:
In the case of `Functor` one needs to be able to go from `F<A> -> F<B>` but that case goes from `F<A> -> G<A>`. One needs the guarantee that those functions return the same container `F<_>`. We currently encode higher kinded parametric typeclasses like this https://github.com/kategory/kategory/blob/master/kategory-core/src/main/kotlin/kategory/typeclasses/Functor.kt

This uses the `HK<F, A>` trick to emulate kinds. which requires later safe downcasts via `ev()` to get from a `HK<Option, A> -> Option<A>`. This is actually what our users complain the most. Also I believe most useful type classes in the domain of FP are higher kinded and many others in OOP can be too. For example write once the Factory of Factories for all `Repository<A>` types that adhere to `Creatable` where `A` is a model class

The biggest issue with client code doing something like `class List<A> : Arg<A>` is that when you write type class instances you actually are doing it over types you can't change. `List` being one of them since it's in the std lib.

The big feature here is to add behaviors to types you have not written yourself without subtyping just by evidence of imports in scope. For example type classes in Kotlin would allow me to add a `Functor` instance for `Deferred` automatically activating:
```
async { doStuff }.map { .. }
```
Now Kotlin Coroutines can become a first class citizen in apps that use Kategory and we can automatically prove laws and provide not just useful combinators but to write polymorphic programs that are independent of the data type: Another example `Async`
```
suspend fun <F<_>, A> bindAsync(fa: F<A>): A given Async<F> = ...
class DeferredAsync : Async<Deferred> {}
class CompletableFutureAsync : Async<CompletableFuture> {}

bindAsync(deferred) // works
bindAsync(future) // works
```
We can effectively abstract away the data types and code to behaviors as you do with interfaces today but no need for subtyping or inheritance in the mix. Also the compiler checks if the dependencies are not in scope. IMO without Higher Kinds we increase the complexity of the encoding and the burden on the user of further passing context around as type args to achieve parametrization over a container type.

Reply:
I'd like to point out that we are usually considering primarily the use cases outside of traditional paradigms (like FP), and more around pragmatic use cases, so we'll need to work on these

Reply:
Classic example where HKT would've been useful for the higher order functions on Iterables (e.g. `map`, `filter`).

Right now, it doesn't matter what type of Collection you have, you'll always get `ArrayList<R>` back. With HKT's you wouldn't have to map over a Set and then convert it back to a set.

Reply:
agreed. I find type classes useful in all contexts. For example here is something library authors can provide with type classes and kinds both in OOP and FP.

Allow users to choose their own return type: 

There are libs that use `async`, others use `Deferred<A>` others use `CompletableFuture<A>`, and the list of runtime possible async containers goes on and on. The lib author can write its libs in terms of `Async` and provide a method such as that:
```
fun <F<_>, A> run(f: () -> A): F<A> given Async<F> = F.runBlock(f)
```
The lib author can now provide in separate modules `lib-deferred` `lib-futures` the following instances in their modules respectively:
```
extension class AsyncDeferred<A>: Async<A> { ... }
extension class CompletableFuture<A>: Async<A> { ... }
```
Now users can choose how to run their programs because they are parametric and can choose their ow return types which works with their current stack. Some users may import the runtime for `CompletableFuture` and others the one for `Coroutines`.
If a user was working for example with `AsyncTask` on Android he could write his own instance of `Async<AsyncTask>` and use the library. Type classes open the option to do ad-hoc polymorphism.

A similar use case is DI, Encoder/Decoders, and the list goes on, all of them verified at compile time for all paradigms.